### ISSUE TITLE
Notifications from the extension only work reliably when the app isn't open

### ISSUE THREAD

**Issue by cotman** [twofas/2fas-ios#81]
Not sure if you'd consider this an app or an extension issue...

I'm using app version 5.2.6 and extension version 1.6.2 (running in Chrome Version 121.0.6167.184).

When I launch the (paired) extension to contact my phone for a code, the notification is nearly always received and displayed if the app isn't running or in the foreground.  However, if the app is already "up" on my phone and displayed, I seldom get a notification from the extension - the "Notification" section in the app is also empty when this happens.  If I close the app and re-click from the extension. then I get the notification and everything works as it should...

**Comment by ghymacx** [comment]
Same here. No push notifications safari extension and iOS 17.4 latest beta.

**Comment by cotman** [comment]
This is oddly variable for me, although not in a good way ;) .  I'd say that around 95% of the time I don't receive notifications from the app, via the extension, if the app is already open on my phone.  However, around 5% of the time I do.

Should have mentioned I'm running iOS 17.3.1.

**Comment by yiveynod** [comment]
I second this issue. I haven’t done any closer investigation on it, but for me I almost never receive a push if the app is already open. The safest way for me to know that I’ll receive the push is to lock my screen.
iOS 17.4

**Comment by enchained** [comment]
I'm using the app (v5.3.5) on iOS 16.6.1, extension 1.6.4 on Edge 122.0.2365.92, and I've had a similar issue. I've enabled the browser extension for the first time, and there were no notifications while the app was open or in the background. Then I force-closed it and clicked the extension once more. Still nothing happened. Then I opened the app manually and I think it automatically offered me to select a service, twice in a row. 

Then I logged out of service and tried for it again, it worked both with closed and running in background. When in foreground, it did open the prompt, but there was no push sound or notification left. I tried the same with a new service, still was not able to clearly reproduce the first issue.

**Comment by adocyn** [comment]
Hi, we've run a battery of tests and we couldn't find anything wrong. To summarise the behaviour is as follows:

- if the app isn't running and device receives push, then it's visible on the notifications screen. User can tap it to open the app or long press to use the notification extension
- if user taps it, then the app will run. Upon every start of the app it checks with the backend (if BE is registered) is there are any awaiting authorisations to be handled
- those authorisations have a time to live. If I remember correctly it's ~ 5 minutes (could be less. 3?). So there could be a situation where there's is a notification on the locked screen, but the app won't ask for authorisation, because all requests have timed out
- if the app is running, upon entering the Tokens tab, the app will pull a list of awaiting authorisations from backend
- if the app is running on the Tokens tap and it received the push notification, it will make the pull on its own and show the authorisation (list or question) screen. There'll be no visible push (sound, Push Notification at the top of the screen), as the app "consumes" it and triggers the pull.
- Push Notifications don't always come immediately. There could be (is) a significant delay. That's why sometimes going into the Tokens tab will trigger fetch quicker than the push.

We monitoring the BE and we'll continue to do so, but for now we can't detect anything wrong with it. If this will persist I will ask you to join our discord and I will send you in private a request for apps logs. From that we can learn more.

**Comment by cotman** [comment]
So, I've installed the latest version of the app, 5.3.5, and I'm now using extension 1.6.4 with Chrome 123.0.6312.86 (on Linux) and now iOS 17.4.1.

Here's what repeats the behaviour for me nearly every time (with an interesting twist, that I've just discovered, at the end :) ):

1. My phone is locked, and 2FAS isn't running
2. I trigger the extension in Chrome
3. I receive a notification on my lock screen which I tap
4. 2FAS starts and provides me with the "Approve" dialogue - the code is filled, all is good :heavy_check_mark: 
5. My phone locks itself
6.  I trigger the extension in Chrome
7. I receive a notification on my lock screen which I tap
8.  2FAS opens but **doesn't** give me an "Approve" dialogue
9. I wait a bit - nothing happens
10. In 2FAS I tap on "Settings", then back to "Tokens", _then_ the  "Approve" dialogue appears...

**Comment by cotman** [comment]
Just as a follow up on my last comment: I've been step 10 as a workaround for the last two weeks - it works 99% of the time.

The issue, at least, for me, is that the "Approve" dialogue needs to be coaxed into making an appearance in the app, and tapping between "Settings" and "Tokens" seems to do that.

**Comment by adocyn** [comment]
@cotman yes, it triggers fetching list of awaiting authorisations from server. I'll check the steps you've provided. Thanks!

**Comment by mbking1** [comment]
I'm seeing the same behavior as @cotman and I use the same solution: click on Settings and then Tokens.  This causes the approval dialogue to appear.

**Comment by adocyn** [comment]
Is this still an issue with newest release?

**Comment by mbking1** [comment]
Yes, this issue has been fixed.  Thank you.

Mark

**Comment by cotman** [comment]
It hasn't been fixed for me - I'm running 5.3.9 and still seem the same behaviour as originally reported.

**Comment by cotman** [comment]
As a chaser to my response above - I've now been using the new version of the app for some time (i.e. 5.3.9) on iOS 17.6.1, and I don't see any change in behaviour from the original bug report - I'm still using the same workaround of tapping "Settings" then "Tokens" to get notifications to appear when the app is already running.

**Comment by ShawnCBerg** [comment]
Anecdotally, I had been experiencing this same behavior with the Chrome extension and iOS app for quite some time, but the last week it seems to not be happening anymore. Came here looking to see if I could find a bug resolution and stumbled on this issue. Not saying it's resolved for everyone, but it hasn't happened to me in about a week and it used to happen once or twice a day before when the Approval dialog wouldn't appear.

**Comment by c0d3h4x0r** [comment]
Even with the latest 2FAS for iPhone, I am experiencing this issue. Notifications appear on the phone, but when I tap on a notification to go into the app, the app does not show the notification... UNLESS I switch to "Settings" (the little gear icon along the bottom edge of the app) and then switch back to "Tokens" (the little padlocked list icon along the bottom edge of the app)... and THEN 2FAS finally shows the in-app notification as expected.

**Comment by cotman** [comment]
As the OP on this, I'm guessing that it's only a minority of iOS users that see this.
I've had this issue since day one, when I moved to 2FAS from Authy; I'm on latest of everything right now (app, extension and iOS version), but still see it.
The "Settings" -> "Token" workaround is almost second nature to me now, which is annoying as it does work properly _sometimes_  - but, as mentioned in 2024, that's typically when the app isn't already open on my iPhone.
It's a shame as 2FAS (aside from this issue)  is hands-down my favourite 2FA app.

**Comment by cotman** [comment]
> I can't recreate that. The Tokens view, when becoming visible, loads the awaiting list of authorizations. I've tried to load the app from scratch, load the app when it's in the background, with or without the PIN. Still - the app loads the list correctly. Are there any particular scenarios you can mention where it doesn't work? Maybe a "always show keyboard" is on?

I'm not sure what more context to give, really, but I'm open to offers 😄 - I don't know what "always show keyboard" is, so I guess I don't have that enabled.

In terms of scenario, it's exactly as listed [here](https://github.com/twofas/2fas-ios/issues/81#issuecomment-2024917763) - in fact, I just had this happen when I logged into GitHub to make this reply:  I'd previously logged into something for work, got the notification and "Approve" dialogue, everything worked as it should - as a result, 2FAS was "up" and running on my phone, which then locked itself.  When I attempted to login to GitHub though,  I got the lock screen notification but no "Approve" dialogue when 2FAS comes up after selecting it  - I had to switch to "Settings" and back to "Tokens" in 2FAS to get it to appear.

**Comment by cotman** [comment]
One further thing on this, that might be helpful (and after re-reading I realise isn't absolutely explicit above): this behaviour only happens, at least for me, if the subsequent launches of 2FAS are made after the phone has locked itself.

So: initially my phone is locked, I trigger the web extension for a code, I get the lock screen notification, tap on it, unlock, 2FAS launches and the "Approve" dialogue comes up fine.  Immediately after that, if the phone has locked itself (which it does automatically after 30 seconds) and I go through the same process, the "Approve" dialogue doesn't appear.
However, if I instead go for another code via the web extension, before the phone has had time to lock, the "Approve" dialogue _does_ come up correctly every time.  So, initial launch from locked is always fine, subsequent launches from locked aren't.

Obviously though, with 30 seconds lock, I'm forever using the "tap 'Settings' then back to 'Tokens'" workaround to get the "Approve" dialogue to appear when I'm normally using 2FAS.

**Comment by Matthew1471** [comment]
Spent ages trying to work out what caused this then found this post.

It's if the phone app is open and it's allowed to go the lock screen. If you force close and re-open the app it'll work again.

**Comment by adocyn** [comment]
I think I have a fix for 6.0 🙂

**Comment by adocyn** [comment]
6.0.0 released.

**Comment by cotman** [comment]
Just a note (as the OP) to say this does now seem to be fixed! 👍
